### PR TITLE
Feature/issue70 code refactoring and cleanup

### PR DESCRIPTION
**Pull request by chenziliang**
Why this PR:
- There has too many concepts in previous code base, splunk sink, splunk logging, splunk client, firehosenozzle, firehoseclient, logrus logging, std/stderr logging etc. Fresh users seems easily get lost by these concepts 
- Unit testing coverage is not good and basically some of the packages are hard to unit test
- Some of the implementation are too verbose and not performant 

How to guarantee this PR (big changes) are more robust and having no regressions ?
- Almost all existing unit test cases still hold just with package name changes
- Unit test code coverage are largely improved. **96+ %**
- Several round of integration testings have been executed

### CODE REVIEW

**Comment by chenziliang** [comment]
screen shot 2017-08-06 at 12 02 19 am

**Comment by chenziliang** [comment]
Blue boxes are abstraction and interfaces. Gray ones are some concrete implementations
screen shot 2017-08-06 at 12 28 46 am

**Comment by chenziliang** [comment]
Take a look at the folder structure of this branch will be easier to glance at

**Comment by Dontre-Splunk** [comment]
@chenziliang - wow you didn't mess around. Looks good from my end.

File: cache/no_cache.go [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package cache
+
+type NoCache struct{}
```
Comment (psychickita) [inline cache/no_cache.go]: NoCache? What does that mean?

**Pull request merged**